Add methods to get active bits from QuantumCircuit

mtreinish opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds two new methods to QuantumCircuit to get the indices of active bits in the circuit. There was an existing rust space method to get a tuple of sets of active bits, but that was not used for these new public methods. This is because they use the bit objects instead of the indices, and also we would want a deterministic iteration order and using a PySet precludes this.

Additionally, the measure_active() method is updated to use this new method internally because it was previously converting to a dag first to compute the idle bits and then measuring the qubits outside that set. Using this new method should be more efficient since it avoid the dag creation.

### Details and comments

mtreinish committed:
Add methods to get active bits from QuantumCircuit

This commit adds two new methods to QuantumCircuit to get the indices of
active bits in the circuit. There was an existing rust space method to
get a tuple of sets of active bits, but that was not used for these new
public methods. This is because they use the bit objects instead of the
indices, and also we would want a deterministic iteration order and
using a PySet precludes this.

Additionally, the measure_active() method is updated to use this new
method internally because it was previously converting to a dag first to
compute the idle bits and then measuring the qubits outside that set.
Using this new method should be more efficient since it avoid the dag
creation.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into active-bit-methods

mtreinish committed:
Remove unused import

debasmita2102 commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -896,6 +896,36 @@ impl CircuitData {
+            .copied()
+            .collect();
+        PyList::new(py, qubits.into_iter().map(|x| x.0))
+    }
```
I had a quick question about `active_qubits()`. In the implementation (lines 904-912 
of circuit_data.rs), it iterates through all instructions and collects qubits from 
qargs. Does this mean directives like `barrier()` and `delay()` would mark qubits as 
"active"?

mtreinish replied:
In the current implementation it would mark them as active. I can filter those instructions very easily, I could have seen an argument either way depending on what you were looking for. Maybe I could add a flag to filter them or not?

jakelishman replied:
It's probably best to default to ignoring directives - things like `measure_active` put a barrier on _all_ qubits, and it'd be weird if `qc.active_qubits()` was different by default before and after a call to `measure_active`.

mtreinish replied:
I added an option in: https://github.com/Qiskit/qiskit/pull/15274/commits/0eaefcbc278272d2d91573499cdd6bcd80f829b5 and it defaults to filtering

jakelishman reviewed (commented):
Aside from Debasmita's (very good!) question - it's odd to me that we're adding a Python-space method that deals in terms of indices and not `Qubit` objects, since that's contrary to almost every other part of the Python-space `QuantumCircuit` API.

Is there a motivating reason why bit indices are necessary here?

mtreinish commented:
> Is there a motivating reason why bit indices are necessary here?

Mostly my distaste for working with `Qubit` and `Clbit` objects. I always prefer using indices. But if people think the python space function should return bit objects I can change it.

mtreinish committed:
Merge branch 'main' into active-bit-methods

mtreinish committed:
Merge remote-tracking branch 'origin/main' into active-bit-methods

mtreinish committed:
Add option to filter inactive operations to active_qubits()

This commit adds a new option to the new methods,
filter_inactive_operations, which as the name implies will filter
inactive operations from being considered. This option is True by
default so that if you run .measure_active() is run it will return the
same qubits that measurements were added for.

mtreinish committed:
Merge branch 'main' into active-bit-methods
